Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,528 @@
+
+  @inlinable
+  public func receiveMetadata(_ headers: HPACKHeaders) {
+    self.interceptors.receive(.metadata(headers))
+  }
+
+  @inlinable
+  public func receiveMessage(_ bytes: ByteBuffer) {
+    do {
+      let message = try self.deserializer.deserialize(byteBuffer: bytes)
+      self.interceptors.receive(.message(message))
+    } catch {
+      self.handleError(error)
+    }
+  }
+
+  @inlinable
+  public func receiveEnd() {
+    self.interceptors.receive(.end)
+  }
+
+  @inlinable
+  public func receiveError(_ error: Error) {
+    self.handleError(error)
+    self.finish()
+  }
+
+  @inlinable
+  public func finish() {
+    switch self.state {
+    case .idle:
+      self.interceptors = nil
+      self.state = .completed
+
+    case let .observing(_, _, statusPromise):
+      statusPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+
+    case .completed:
+      self.interceptors = nil
+    }
+  }
+
+  // MARK: - Interceptors to User Function
+
+  @inlinable
+  internal func receiveInterceptedPart(_ part: GRPCServerRequestPart<Request>) {
+    switch part {
+    case let .metadata(headers):
+      self.receiveInterceptedMetadata(headers)
+    case let .message(message):
+      self.receiveInterceptedMessage(message)
+    case .end:
+      self.receiveInterceptedEnd()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedMetadata(_ headers: HPACKHeaders) {
+    switch self.state {
+    case .idle:
+      let statusPromise: EventLoopPromise<GRPCStatus> = context.eventLoop.makePromise()
+
+      // Make a context to invoke the user handler with.
+      let context = GRPCAsyncServerCallContext(
+        headers: headers,
+        logger: self.context.logger,
+        userInfoRef: self.userInfoRef
+      )
+
+      // Create a request stream to pass to the user function and capture the
+      // handler in the updated state to allow us to produce more results.
+      let requestStream = GRPCAsyncStream<Request>(AsyncThrowingStream { continuation in
+        self.state = .observing(
+          { streamEvent in
+            switch streamEvent {
+            case let .message(request): continuation.yield(request)
+            case .end: continuation.finish()
+            }
+          },
+          context,
+          statusPromise
+        )
+      })
+
+      // Create a writer that the user function can use to pass back responses.
+      let responseStreamWriter = GRPCAsyncResponseStreamWriter(
+        context: context,
+        compressionIsEnabled: self.context.encoding.isEnabled,
+        sendResponse: self.interceptResponse(_:metadata:)
+      )
+
+      // Send response headers back via the interceptors.
+      self.interceptors.send(.metadata([:]), promise: nil)
+
+      // Register callbacks on the status future.
+      statusPromise.futureResult.whenComplete(self.userFunctionStatusResolved(_:))
+
+      // Spin up a task to call the async user handler.
+      self.task = statusPromise.completeWithTask {
```

Comment:
We'll need to revisit this, I think. We may not be writable when we complete the user function (which means there may also be buffered response messages to send before the status). Completing the user function should instead enqueue a status to send rather immediately send one back. I suspect we'll need a more detailed `State` enum to represent state as well (I don't have a clear picture what this looks like yet but completing a promise is probably not the right way to go).

Reply:
Is this something we revisit when we have the new writer or something we do in this PR?

Reply:
It'll be clearer when we have the new writer.

Reply:
I think we resolved this now.